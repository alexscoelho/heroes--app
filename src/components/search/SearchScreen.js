import React, { useMemo } from "react";
import { useLocation } from "react-router-dom";
import { HeroeCard } from "../heroes/HeroeCard";
import { useForm } from "../hooks/useForm";
import { getHeroesByName } from "../selectors/getHeroesByName";
const queryString = require("query-string");

export const SearchScreen = ({ history }) => {
  const location = useLocation();
  const { q = "" } = queryString.parse(location.search);

  const [formValues, handleInputChange] = useForm({
    query: q,
  });
  const { query } = formValues;

  const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

  const handleSearch = (e) => {
    e.preventDefault();
    history.push(`?q=${query}`);
  };
  return (
    <div>
      <h1>Search Screen</h1>
      <hr />

      <div className='row'>
        <div className='col-5'>
          <h4>Search Form</h4>
          <hr />

          <form onSubmit={handleSearch}>
            <input
              onChange={handleInputChange}
              type='text'
              name='query'
              placeholder='Find your hero'
              className='form-control'
              value={query}
            />
            <button
              type='submit'
              className='btn m-1 btn-block btn-outline-primary'
            >
              Search..
            </button>
          </form>
        </div>
        <div className='col-7'>
          <h4>Results</h4>
          <hr />

          {q === "" && <div className='alert alert-info'>Search hero</div>}

          {q !== "" && !heroesFiltered.length && (
            <div className='alert alert-danger'>There is no hero with {q} </div>
          )}

          {heroesFiltered.map((heroe) => (
            <HeroeCard key={heroe.id} {...heroe} />
          ))}
        </div>
      </div>
    </div>
  );
};
