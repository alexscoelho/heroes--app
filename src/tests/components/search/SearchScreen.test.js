import { mount } from "enzyme";
import React from "react";
import { MemoryRouter, Route } from "react-router-dom";
import { SearchScreen } from "../../../components/search/SearchScreen";

describe("Pruebas en <SearchScreen/>", () => {
  test("debe de mostrarse correctamente con valores por defecto", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route path='/search' component={SearchScreen} />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".alert-info").text().trim()).toBe("Search hero");
  });
  test("debe de mostrar a batman y el input con el valor del querystring", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=superman"]}>
        <Route path='/search' component={SearchScreen} />
      </MemoryRouter>
    );
    expect(wrapper.find("input").prop("value")).toBe("superman");
    expect(wrapper).toMatchSnapshot();
  });
  test("debe de mostrar un error si no encuentra el hero", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=potato"]}>
        <Route path='/search' component={SearchScreen} />
      </MemoryRouter>
    );
    wrapper.find("button").simulate("click");
    expect(wrapper.find(".alert").text().trim()).toBe(
      "There is no hero with potato"
    );
    expect(wrapper).toMatchSnapshot();
  });
  test("debe de llamar el push del history", () => {
    const history = {
      push: jest.fn(),
    };
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=potato"]}>
        <Route
          path='/search'
          component={() => <SearchScreen history={history} />}
        />
      </MemoryRouter>
    );
    wrapper.find("input").simulate("change", {
      target: {
        name: "query",
        value: "batman",
      },
    });
    wrapper.find("form").prop("onSubmit")({
      preventDefault() {},
    });
    expect(history.push).toHaveBeenCalledWith("?q=batman");
  });
});
