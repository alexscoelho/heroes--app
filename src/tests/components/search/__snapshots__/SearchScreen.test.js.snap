// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen/> debe de mostrar a batman y el input con el valor del querystring 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="query"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="superman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search..
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Superman"
              className="card-img"
              src="/assets/heroes/dc-superman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Superman
              </h5>
              <p
                className="card-text"
              >
                Kal-El
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Action Comics #1
                </small>
              </p>
              <a
                href="/heroe/dc-superman"
                onClick={[Function]}
              >
                More..
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen/> debe de mostrar un error si no encuentra el hero 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="query"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="potato"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search..
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        There is no hero with 
        potato
         
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen/> debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="query"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search..
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search hero
      </div>
    </div>
  </div>
</div>
`;
